File: backend/app/models/post_rating.py
from datetime import datetime
from app import db

class PostRating(db.Model):
    __tablename__ = 'post_ratings'
    
    id = db.Column(db.Integer, primary_key=True)
    post_id = db.Column(db.Integer, db.ForeignKey('posts.id'), nullable=False)
    user_id = db.Column(db.Integer, db.ForeignKey('users.id'), nullable=False)
    rating = db.Column(db.Integer, nullable=False)  # Rating value (e.g., 1-5)
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    updated_at = db.Column(db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # Unique constraint to prevent multiple ratings from the same user
    __table_args__ = (
        db.UniqueConstraint('post_id', 'user_id', name='unique_post_rating'),
    )
    
    def to_dict(self):
        return {
            'id': self.id,
            'post_id': self.post_id,
            'user_id': self.user_id,
            'rating': self.rating,
            'created_at': self.created_at.isoformat(),
            'updated_at': self.updated_at.isoformat()
        }

File: backend/src/models/PostRating.js
const { DataTypes } = require('sequelize');
const sequelize = require('../config/database');
const User = require('./User');
const Post = require('./Post');

const PostRating = sequelize.define('PostRating', {
  id: {
    type: DataTypes.UUID,
    defaultValue: DataTypes.UUIDV4,
    primaryKey: true
  },
  post_id: {
    type: DataTypes.UUID,
    allowNull: false,
    references: {
      model: 'posts',
      key: 'id'
    },
    onDelete: 'CASCADE'
  },
  user_id: {
    type: DataTypes.UUID,
    allowNull: false,
    references: {
      model: 'users',
      key: 'id'
    },
    onDelete: 'CASCADE'
  },
  rating: {
    type: DataTypes.STRING(10),
    allowNull: false,
    // 'left', 'right'
  },
  created_at: {
    type: DataTypes.DATE,
    defaultValue: DataTypes.NOW
  },
  is_rewarded: {
    type: DataTypes.BOOLEAN,
    defaultValue: false
  },
  reward_amount: {
    type: DataTypes.INTEGER,
    allowNull: true
  }
}, {
  tableName: 'post_ratings',
  timestamps: false,
  indexes: [
    {
      unique: true,
      fields: ['post_id', 'user_id']
    }
  ]
});

// Define associations
PostRating.belongsTo(User, { foreignKey: 'user_id' });
User.hasMany(PostRating, { foreignKey: 'user_id' });

PostRating.belongsTo(Post, { foreignKey: 'post_id' });
Post.hasMany(PostRating, { foreignKey: 'post_id' });

module.exports = PostRating;
